### PR TITLE
tools: Add AsyncChecker

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: tools: Add AsyncChecker
Additional Description:

Allows use of asyncio for checkers

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
coverage is here https://storage.googleapis.com/envoy-pr/17172/tooling/index.html

File: tools/base/tests/test_checker.py [hunk 1]
```diff
@@ -654,3 +654,120 @@ def test_bazelchecker_constructor():
...
+            await checker.on_checks_complete()
+            == m_complete.return_value)
+
+    assert (
+        list(m_complete.call_args)
+        == [(), {}])
+
+
+@pytest.mark.asyncio
+async def test_asynchecker_dunder_run(patches):
```
Comment (phlax) [inline tools/base/tests/test_checker.py:723]: change name of this

**Review by htuch — APPROVED** [review]
LGTM, but as discussed on chat, I'd like ensure we generally have the philosophy that we keep tests hermetic that can be reasonably made hermetic, so we should reserve the use of aio based libraries (e.g. HTTP fetch) to just the situations where this is inherently necessary in the check definition.

As a principle, developers should be able to run their checkers for the most part with no runtime (vs. setup) network connectivity I think.

**Comment by phlax** [comment]
in the immediate case the network connectivity is to the docker.sock - which is currently setup if you run with `./ci/run_envoy_docker.sh ...` which will mount the docker.sock into the container - so its an assumption baked into the current local test env

as a general thing - for testing envoy itself its inevitable that we have to use tools on top to communicate with it - i have tried to keep these to fairly standard ~unix utils, and curl, jq etc

in terms of python libs i would generally prefer using asyncio libs over the sync ones - but my initial use for this is more because i was struggling to get docker-compose to build/run on arm64 - so it was easier to do it in python

**Comment by htuch** [comment]
Yeah, HTTP is fine for data plane tests on localhosts (assuming we don't assume anything about specific ports etc.). I was more thinking that the checkers we have here are developer tools for validating things like code format. Is there a plan to evolve this into a more general testing framework?

**Comment by phlax** [comment]
> Is there a plan to evolve this into a more general testing framework?

not really - im trying to keep it reasonably simple - just the runner and checker classes

theres probably some stuff we could re/move to bazel - i did it in python i think because that was how i knew to do it - but mostly what is there is working quite well

the code that we register with `envoy_py_*` gets tested and the coverage is still at 100%

**Comment by phlax** [comment]
holding off on this while we figure out a way forward for #16899

**Comment by phlax** [comment]
on the basis that this is generally useful and will i think be used in #16899 im gonna land it

**Pull request merged**